File: Dynamic Programming/241. Different Ways to Add Parentheses/Solution.py
# https://leetcode.com/problems/different-ways-to-add-parentheses/

# By Jiapeng
# 记忆化递归求解分段问题 partition problem
# Runtime: 32 ms, faster than 85.56% of Python3 online submissions for Different Ways to Add Parentheses.
# Memory Usage: 14.2 MB, less than 5.35% of Python3 online submissions for Different Ways to Add Parentheses.
class Solution:
    def diffWaysToCompute(self, input: str) -> List[int]:
        """
        记忆化递归
        每一段s都可以以一个运算符号为界分成左右两边，左边式一个集合，右边是一个集合
        根据运算符号的不同求左右两边集合对应的笛卡尔积集合
        for循环中计算出的所有笛卡尔积集合做并集，得出当前递归的结果并返回
        同时把返回值都存在dict里面方便再用的时候查找（记忆化）
        当发现当前递归求出的笛卡尔积集合是空集的时候，说明该段input无运算符号，已经分解至只剩数字，此时直接返回带有该数字的list即可
        """
        def ways(s):
            if array[s]:
                return array[s]
            cur = []
            for i in range(len(s)):
                if s[i] in '+-*':
                    left = ways(s[:i])
                    right = ways(s[i+1:])
                    for l in left:
                        for r in right:
                            cur.append(opts[s[i]](l, r))
            if not cur:
                cur.append(int(s))
            array[s] = cur
            return cur
        
        array = collections.defaultdict(list)
        opts = {'+': lambda a, b: a + b,
                '-': lambda a, b: a - b,
                '*': lambda a, b: a * b}
        return ways(input)
    
       

File: Dynamic Programming/241. Different Ways to Add Parentheses/Solution.java
// https://leetcode.com/problems/different-ways-to-add-parentheses/

/**
By Jiapeng
记忆化递归求解分段问题
Runtime: 1 ms, faster than 99.68% of Java online submissions for Different Ways to Add Parentheses.
Memory Usage: 38.3 MB, less than 85.09% of Java online submissions for Different Ways to Add Parentheses.
**/
class Solution {
    private static Map<String, List<Integer>> arrays = new HashMap<>();     // 记忆化每次递归结果
    
    public List<Integer> diffWaysToCompute(String input) {
        if(arrays.containsKey(input))
            return arrays.get(input);
        List<Integer> cur = new ArrayList<>();
        for(int i = 0; i < input.length(); i++){
            char ch = input.charAt(i);
            if(ch == '+' || ch == '-' || ch == '*'){
                for(Integer l: diffWaysToCompute(input.substring(0, i)))
                    for(Integer r: diffWaysToCompute(input.substring(i + 1))){
                        if(ch == '+') cur.add(l + r);
                        else if (ch == '-') cur.add(l - r);
                        else cur.add(l * r);
                    }
            }
        }
        if(cur.isEmpty())       // cur为空说明当前input不存在'+','-','*'符号
            cur.add(Integer.parseInt(input));
        arrays.put(input, cur);
        return cur;
    }
}

/**
By Jiapeng
和方法1一模一样，但是使用了传递函数的高级用法！！！
Runtime: 2 ms, faster than 63.20% of Java online submissions for Different Ways to Add Parentheses.
Memory Usage: 37.7 MB, less than 94.37% of Java online submissions for Different Ways to Add Parentheses.
**/
class Solution2 {
    private static Map<String, List<Integer>> arrays = new HashMap<>();
    private static Map<Character, BiFunction<Integer, Integer, Integer>> ops = new HashMap<>();     // Map存储对应的函数，详情参见java8函数式接口
    {                                                                                               // https://www.runoob.com/java/java8-functional-interfaces.html
        ops.put('+', (a, b) -> a + b);
        ops.put('-', (a, b) -> a - b);
        ops.put('*', (a, b) -> a * b);
    }
    public static Integer calc(BiFunction<Integer, Integer, Integer> bi, Integer l, Integer r) {    // 使用apply() call function
        return bi.apply(l, r);
    }
    public List<Integer> diffWaysToCompute(String input) {
        if(arrays.containsKey(input))
            return arrays.get(input);
        List<Integer> cur = new ArrayList<>();
        for(int i = 0; i < input.length(); i++){
            char ch = input.charAt(i);
            if(ch == '+' || ch == '-' || ch == '*'){
                for(Integer l: diffWaysToCompute(input.substring(0, i)))
                    for(Integer r: diffWaysToCompute(input.substring(i + 1))){
                        cur.add(calc(ops.get(ch), l, r));
                    }
            }
        }
        if(cur.isEmpty())
            cur.add(Integer.parseInt(input));
        arrays.put(input, cur);
        return cur;
    }
}
